include/deal.II/numerics/smoothness_estimator.h

```diff
@@ -0,0 +1,385 @@
+                             Vector<float> &smoothness_indicators,
+                             const std::function<void(std::vector<bool> &flags)>
+                               &coefficients_predicate =
+                                 [](std::vector<bool> &flags) -> void {
+                               std::fill(flags.begin(), flags.end(), true);
+                             },
+                             const double smallest_abs_coefficient = 1e-10,
+                             const bool   only_flagged_cells       = false);
+
+  /**
+   * Same as the function above, but with a default configuration for the
+   * Legendre series expansion FESeries::Legendre.
+   *
+   * We use as many modes as the highest polynomial of all finite elements used
+   * increased by one, since we start with the first Legendre polynomial which
+   * is just a constant. Further, we use a Gaussian quadrature designed to yield
+   * exact results for the highest order Legendre polynomial used.
+   */
+  template <int dim, int spacedim, typename VectorType>
+  void
+  legendre_coefficient_decay(
+    const hp::DoFHandler<dim, spacedim> &  dof_handler,
+    const std::vector<const VectorType *> &all_solutions,
+    std::vector<Vector<float> *> &         all_smoothness_indicators,
+    const std::function<void(std::vector<bool> &flags)>
+      &coefficients_predicate = [](std::vector<bool> &flags) -> void {
+      std::fill(flags.begin(), flags.end(), true);
+    },
+    const double smallest_abs_coefficient = 1e-10,
+    const bool   only_flagged_cells       = false);
+
+  /**
+   * Same as the function above, only for one @p solution vector.
+   */
+  template <int dim, int spacedim, typename VectorType>
+  void
+  legendre_coefficient_decay(const hp::DoFHandler<dim, spacedim> &dof_handler,
+                             const VectorType &                   solution,
+                             Vector<float> &smoothness_indicators,
+                             const std::function<void(std::vector<bool> &flags)>
+                               &coefficients_predicate =
+                                 [](std::vector<bool> &flags) -> void {
+                               std::fill(flags.begin(), flags.end(), true);
+                             },
+                             const double smallest_abs_coefficient = 1e-10,
+                             const bool   only_flagged_cells       = false);
+
+  /**
+   * Estimate the smoothness of a solution based on the decay of coefficients
+   * from a Fourier series expansion.
+   *
+   * From the definition, we can write our Fourier series expansion
+   * $\hat U_{\bf k}$ as a matrix product
+   * @f[
+   *    \hat U_{\bf k}
+   *    = {\cal F}_{{\bf k},j} u_j,
+   * @f]
+   * with $u_j$ the coefficients and ${\cal F}_{{\bf k},j}$ the transformation
+   * matrix from the expansion. We use the class FESeries::Fourier to determine
+   * all coefficients $u_j$.
```

Comment:
Did you mean `U_j` here?

Reply:
I guess in this case, it even has to be `U_{\bf k}`.